### PR TITLE
New merge branch

### PR DESCRIPTION
**Pull request by thomas-waite**
Implemented the Bilateral Swap feature, which enables a bilateral swap in zero knowledge using AZTEC. 

The implementation consists of three main components:
1) Proof verification smart contract - BilateralSwap.sol 
2) Ganache/Truffle testing script - bilateralSwap.js

( 3) Proof construction Javascript code - believe it has already been merged/will be)

### CODE REVIEW

File: contracts/AZTEC/BilateralSwap.sol [hunk 1]
```diff
@@ -0,0 +1,249 @@
+pragma solidity ^0.4.24;
```
Comment (zac-williamson) [inline contracts/AZTEC/BilateralSwap.sol]: We should update this file to pass the new Solium linter specifications

File: contracts/AZTEC/BilateralSwap.sol [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+                    /*
+                    Storing 128 bytes from position index in call data to the position in memory where the notes hashed commitments 
+                    should be stored
+
+                    Think the commitments are 128 bytes big
+                    */
+
+                    calldatacopy(add(0x300, mul(i, 0x80)), index, 0x80)
+                }
+                // storing at position 0x00 in memory, the kecca hash of everything from start of the commitments to the end
```
Comment (zac-williamson) [inline contracts/AZTEC/BilateralSwap.sol]: keccak

File: test/BilateralSwap/bilateralSwap.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            const fakeProofData = new Array(4).map(() => new Array(6).map(() => toBytes32.randomBytes32()));
+
+            await exceptions.catchRevert(BilateralSwap.validateBilateralSwap(fakeProofData, challenge, t2, {
+                from: accounts[0],
+                gas: 4000000,
+            }));
+        });
+
+        it('Validate failure when points not on curve', async () => {
+            const zeroes = `${padLeft('0', 64)}`;
```
Comment (zac-williamson) [inline test/BilateralSwap/bilateralSwap.js]: We should update this to produce a satisfying proof *if* we don't check for whether points are on the curve.

i.e. the x and y coordinates for the 4 notes should be '0', but the challenge should correctly resolve to the hash of the commitments and blinding factors (also points with x/y coordinates of '0')

File: test/BilateralSwap/bilateralSwap.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            testNotes = helpers.makeTestNotes(makerNoteValues, takerNoteValues);
+        });
+
+        it('validate that the Javascript proof is constructed correctly', () => {
+            const { proofData, challenge } = bilateralProof.constructBilateralSwap(testNotes, accounts[0]);
+            const result = bilateralProof.verifyBilateralSwap(proofData, challenge, accounts[0]);
+            expect(result).to.equal(true);
+        });
+
+        it('validate that the smart contract can verify the bilateral swap proof', async () => {
```
Comment (zac-williamson) [inline test/BilateralSwap/bilateralSwap.js]: we should add test cases for where the a bid/ask note pair has a value of '0', and validate that the proof passes

**Comment by PaulRBerg** [comment]
Closed, migrated to the monorepo in the meantime.

**Pull request closed without merging**